internal/locate/region_cache.go

```diff
@@ -1203,6 +1204,160 @@ func (c *RegionCache) LocateKeyRange(bo *retry.Backoffer, startKey, endKey []byt
+
+	// 2. load remaining regions from pd client
+	for len(uncachedRanges) > 0 {
+		regions, err := c.BatchLoadRegionsWithKeyRanges(bo, uncachedRanges, defaultRegionsPerBatch, opts...)
+		if err != nil {
+			return nil, err
+		}
+		if len(regions) == 0 {
+			return nil, errors.Errorf("BatchLoadRegionsWithKeyRanges return empty batchedRegions without err")
+		}
+
+		for _, r := range regions {
+			merger.appendRegion(r)
+		}
+		// if the regions are not loaded completely, split uncachedRanges and load the rest.
+		uncachedRanges = nextRanges(uncachedRanges, regions[len(regions)-1].EndKey())
+	}
+	return merger.build(), nil
+}
+
+type batchLocateRangesMerger struct {
+	lastEndKey      util.Option[[]byte]
+	cachedIdx       int
+	cachedRegions   []*Region
+	mergedLocations []*KeyLocation
+}
+
+func newBatchLocateRegionMerger(cachedRegions []*Region, sizeHint int) *batchLocateRangesMerger {
+	return &batchLocateRangesMerger{
+		lastEndKey:      util.None[[]byte](),
+		cachedRegions:   cachedRegions,
+		mergedLocations: make([]*KeyLocation, 0, sizeHint),
+	}
+}
+
+func (m *batchLocateRangesMerger) appendKeyLocation(r *Region) {
+	m.mergedLocations = append(m.mergedLocations, &KeyLocation{
+		Region:   r.VerID(),
+		StartKey: r.StartKey(),
+		EndKey:   r.EndKey(),
+		Buckets:  r.getStore().buckets,
+	})
+}
+
+func (m *batchLocateRangesMerger) appendRegion(uncachedRegion *Region) {
+	defer func() {
+		endKey := uncachedRegion.EndKey()
+		if len(endKey) == 0 {
+			m.cachedIdx = len(m.cachedRegions)
+		} else {
+			m.lastEndKey = util.Some(uncachedRegion.EndKey())
+		}
+	}()
+	if len(uncachedRegion.StartKey()) == 0 {
+		m.appendKeyLocation(uncachedRegion)
+		return
+	}
+	lastEndKey := m.lastEndKey.Inner()
+	if lastEndKey != nil && bytes.Compare(*lastEndKey, uncachedRegion.StartKey()) >= 0 {
+		// the uncached regions are continued, do not consider cached region by now.
+		m.appendKeyLocation(uncachedRegion)
+		return
+	}
+	for ; m.cachedIdx < len(m.cachedRegions); m.cachedIdx++ {
+		if lastEndKey != nil && bytes.Compare(*lastEndKey, m.cachedRegions[m.cachedIdx].EndKey()) >= 0 {
+			// skip the cached region that is covered by the uncached region.
+			continue
+		}
+		if bytes.Compare(m.cachedRegions[m.cachedIdx].StartKey(), uncachedRegion.StartKey()) >= 0 {
+			break
+		}
+		m.appendKeyLocation(m.cachedRegions[m.cachedIdx])
+	}
+	m.appendKeyLocation(uncachedRegion)
+}
+
+func (m *batchLocateRangesMerger) build() []*KeyLocation {
+	// append the rest cache hit regions
+	for ; m.cachedIdx < len(m.cachedRegions); m.cachedIdx++ {
+		if lastEndKey := m.lastEndKey.Inner(); lastEndKey != nil && bytes.Compare(*lastEndKey, m.cachedRegions[m.cachedIdx].EndKey()) >= 0 {
+			// skip the cached region that is covered by the uncached region.
+			continue
+		}
+		m.appendKeyLocation(m.cachedRegions[m.cachedIdx])
+	}
+	return m.mergedLocations
+}
+
+func nextRanges(keyRanges []pd.KeyRange, locatedKey []byte) []pd.KeyRange {
```

Comment:
Better to explain the input parameters for this function, also the returned slice is referenced but not cloned. The `nextRanges` is more understandable as a local function or closure, as a standalone function how about `RangesAfterKey`.

Reply:
A closure can not be tested independently, let me rename and comment it.